pkg/autoscaler/metrics/http_scrape_client.go

```diff
@@ -32,6 +33,54 @@ type httpScrapeClient struct {
 	httpClient *http.Client
 }
 
+var step = 64
+
+var dataSizeClasses = []int{
+	step,
+	step * 2,
+	step * 3,
+	step * 4,
+	304, // 292 is the maximum proto buf stat record, use multiple of 16
+}
+
+var pools = [...]sync.Pool{
+	{
+		New: func() interface{} {
+			return make([]byte, dataSizeClasses[0], dataSizeClasses[0])
+		},
+	},
+	{
+		New: func() interface{} {
+			return make([]byte, dataSizeClasses[1], dataSizeClasses[1])
+		},
+	},
+	{
+		New: func() interface{} {
+			return make([]byte, dataSizeClasses[2], dataSizeClasses[2])
+		},
+	},
+	{
+		New: func() interface{} {
+			return make([]byte, dataSizeClasses[3], dataSizeClasses[3])
+		},
+	},
+	{
+		New: func() interface{} {
+			return make([]byte, dataSizeClasses[4], dataSizeClasses[4])
+		},
+	},
+}
+
+func getDataBuffer(size int) ([]byte, int) {
+	i := 0
+	for ; i < len(dataSizeClasses)-1; i++ {
+		if size <= dataSizeClasses[i] {
+			break
+		}
+	}
+	return pools[i].Get().([]byte), i
```

Comment:
just add a sentinel larger value and fold `i` into the loop and directly return from there.
And then panic in the end.